Feature: Transition to celery for background tasks

stumpylog opened the pull request:
## Proposed change

With `django-q` remaining unmaintained and myself being the main keeper of our own copy of it, it is probably best to transition away from it and to something supported, documented, etc.

This PR transitions the backend to use [celery](https://docs.celeryq.dev/en/stable/index.html) and celery beat for background tasks and scheduled tasks.  It is maintained, has extensive documentation, at least some Django integrations and generally a lot of features.  

I have successfully migrated a 1.8.0 Docker instance to this image, so I don't believe this is a breaking change.  For bare metal installs, it will require some more work, but I believe that is already the case.  All the tasks, including scheduled tasks are working.  The frontend task view is not yet working

### TODOs
- [x] Get frontend File Tasks view working
- [x] Fix frontend testing
- [x] [Update documentation](https://github.com/search?q=qcluster+repo%3Apaperless-ngx%2Fpaperless-ngx+filename%3A*.rst&type=Code&ref=advsearch&l=&l=)
- [x] Extensive testing

### Testing
- Stress Test
  - Consumed ~1000 files
  - postgres as the db
  - 16 workers.
  - All files were from an export
  - All files consumed, frontend tasks updated as expected
- In-place Upgrade
  - 1.7 -> this branch, no issues
  - 1.8 -> this branch, no issues
  - 1.9 -> this branch, no issues

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

stumpylog committed:
Transitions the backend to celery and celery beat

shamoon commented:
Do you think its possible to migrate the frontend task queue without that unmerged PR you linked? I dont pretend to know the library off-hand but I see celery [makes good use of signals](https://docs.celeryq.dev/en/stable/userguide/signals.html) so could we transition the q signals to those?

stumpylog commented:
It could be, but I was considering the number of odd database problems we had around using signals and not really wanting to have those happening again.  If that PR gets merged, then the code for creating a model in the right times is in django-celery-results, while we would only need to extend it with the `acknowledged` field.  Basically, less for us to manage and do.

It's possible celery is better behaved with database connections and signals, so I'll look into it more.  

The view to tasks is the last large part still TODO.  Once I have something in the backend, I'll need your expert skills to fix my damage to the frontend.

shamoon commented:
Gotcha. Of course I would be happy to help, feel free to break the frontend. I will take a look at celery's signal stuff when I have some time, it took me a while to figure it out with q so I have a feeling you (or someone else) will get there first. Either way obviously I defer to you on the best way to go about it.

stumpylog committed:
Finalizes what the PaperlessTask will look like to the frontend

stumpylog commented:
I hooked into the TaskResult post_save signal instead.  It's the closest thing to abstract away a lot of the when to save/update our model.  The JSON looks pretty close now, but there's a few differences, as the fields were directly one to one.

- `created` is `date_created`
- No real equivalent to `started`, closest is `date_done`
- Status is now one of `{'STARTED', 'RETRY', 'SUCCESS', 'REVOKED', 'RECEIVED', 'PENDING', 'FAILURE'}`
  - if it's helpful, the backend could probably condense some of those

So maybe when you've got some time, you could work on the frontend portion needed for the differences

shamoon commented:
That was fast! Ok I'll take a look, one thing as I was just firing things up to poke around is when I try to migrate my test install I think the migrations dont work if `attempted_task` isnt null, e.g. Im getting:
```
django.db.utils.IntegrityError: The row in table 'documents_paperlesstask' with primary key '11' has an invalid foreign key: documents_paperlesstask.attempted_task_id contains a value '5f8e8ed5e36242639f52405e5a034803' that does not have a corresponding value in django_q_task.id
```
presumably because we dropped the q tables already. Do we want to null-out those values first?

stumpylog commented:
Oh, I see.  I re-ordered the migrations, but I forgot I did that order for a reason.  One sec

stumpylog committed:
Fixes migration order back to the right way

stumpylog committed:
Ensures all existing one to one fields are nulled before altering the field

shamoon committed:
Fix tab ids & date parsing

shamoon committed:
Add document name & error result parsing to PaperlessTask serializer

shamoon committed:
Filter out old migrated tasks

shamoon committed:
Add related_document and direct link from task UI

shamoon commented:
Ok I think I got the frontend part squared away! I also took the opportunity to add a little feature where we link directly to a doc if its a successful consumption, e.g.:
![related](https://user-images.githubusercontent.com/4887959/192683665-30aca5e3-a98a-4ae8-b58a-ef691c37e13f.jpg)

Of course feel free to modify what I did, I did have to do some funny things like wrangle the `task_kwargs` content into a proper JSON object string, etc.

shamoon committed:
re-enable frontend tests and fix tasks fixture

stumpylog committed:
Updates how task_args and task_kwargs are parsed, adds testing to cover everything I can think of

stumpylog commented:
Yeah, that's some weird storage of task_args and task_kwargs.  It's like a double string of a Python container.  I found a few open issues in celery and celery-results about it, but nothing merged or fixed.  I changed it up a little bit and added testing to cover the cases I could think of.

The document open is a nice touch.

shamoon commented on src/documents/serialisers.py:
```diff
@@ -632,24 +652,108 @@ def get_type(self, obj):
+                    result = filepath.name
+            except Exception as e:  # pragma: no cover
+                # Extra security if something is malformed
+                logger.warn(f"Error getting task result: {e}", exc_info=True)
```
```suggestion
                logger.warn(f"Error getting file name from task: {e}", exc_info=True)
```

=)

stumpylog committed:
Fix logger location tag

Co-authored-by: shamoon <4887959+shamoon@users.noreply.github.com>

stumpylog committed:
Cleans up and documnets installation on bare metal, includes systemd service file for the celery and celery beat

stumpylog commented:
Alright, I've tweaked the documentation and added the necessary systemd fie.  I want to test out more just to be sure about upgrading in place (migrations in particular), but I think this is fully ready now.

shamoon committed:
Update references to qcluster t celery

stumpylog commented:
This will definitely fix #1321 and I hope fix #1355

stumpylog committed:
Fixes up some issues with the migrations and type mismatches

stumpylog committed:
Fixes django-q still in the Pipfile.lock

shamoon reviewed (approved):
Not sure if the current test failures are real (they look weird, Im just re-running) but I think this is great and I vote we forge ahead as discussed in the team thread as well

shamoon commented:
Ah
```
Error: Unable to locate the previous sha: 71b45715244dd036948e7cf38307e7f29c7142db
[252](https://github.com/paperless-ngx/paperless-ngx/actions/runs/3177304226/jobs/5179989679#step:9:256)
  Error: You seem to be missing 'fetch-depth: 0' or 'fetch-depth: 2'. See https://github.com/tj-actions/changed-files#usage
[253](https://github.com/paperless-ngx/paperless-ngx/actions/runs/3177304226/jobs/5179989679#step:9:257)
  Error: Process completed with exit code 1.
```
maybe from #1702?

stumpylog commented:
They're not really. The tests pass, it's a problem with the changed files action.  Don't think it likes me rebasing

shamoon committed:
Merge branch 'dev' into feature-use-celery

shamoon commented:
Passing now 😎

stumpylog committed:
Fixes migration error if some tasks are defined already

shamoon commented:
Alright lets do it, clearly next release will need some good beta testing =)
